staging_service/import_specifications/individual_parsers.py

```diff
@@ -0,0 +1,164 @@
+
+_DATA_TYPE = "Data type:"
+_VERSION_STR = "Version:"
+_HEADER_SEP = ";"
+_EXPECTED_HEADER = f"{_DATA_TYPE} <data_type>{_HEADER_SEP} {_VERSION_STR} <version>"
+_HEADER_REGEX = re.compile(f"{_DATA_TYPE} (\\w+){_HEADER_SEP} {_VERSION_STR} (\\d+)")
+
+
+class _ParseException(Exception):
+    pass
+
+
+def _parse_header(line: str, spec_source: SpecificationSource, maximum_version: int) -> str:
+    # return is data type
+    match = _HEADER_REGEX.fullmatch(line)
+    if not match:
+        raise _ParseException(Error(
+            ErrorType.PARSE_FAIL,
+            f'Invalid header; got "{line}", expected "{_EXPECTED_HEADER}"',
+            spec_source
+        ))
+    version = int(match[2])
+    if version > maximum_version:
+        raise _ParseException(Error(
+            ErrorType.PARSE_FAIL,
+            f"Schema version {version} is larger than maximum processable "
+            + f"version {maximum_version}",
+            spec_source
+        ))
+    return match[1]
+
+
+def _get_datatype(input_: TextIO, spec_source: SpecificationSource, maximum_version: int) -> str:
+    try:
+        return _parse_header(next(input_).strip(), spec_source, maximum_version)
+    except StopIteration:
+        raise _ParseException(Error(
+            ErrorType.PARSE_FAIL, "Missing data type / version header", spec_source))
+
+
+def _error(error: Error) -> ParseResults:
+    return ParseResults(errors = tuple([error]))
+
+
+def _strip(putative_str: PRIMITIVE_TYPE) -> PRIMITIVE_TYPE:
+    return putative_str.strip() if isinstance(putative_str, str) else putative_str
+
+
+def _check_for_duplicate_headers(headers: pandas.Index, spec_source: SpecificationSource):
+    seen = set()
+    for name in headers.get_level_values(0):
+        if name in seen:
+            raise _ParseException(Error(
+                ErrorType.PARSE_FAIL,
+                f"Duplicate column name: {name.split('.')[0]}",
+                spec_source
+            ))
+        seen.add(name)
+
+
+def _check_xsv_row_count(path: Path, expected_count: int, sep: str):
+    with open(path) as input_:
+        # since we parsed the first line in the main _parse_xsv method, just discard here
+        next(input_)
+        for i, line in enumerate(input_):
+            # I suppose we could count-- if the last entry is whitespace but f it
+            count = 0 if not line.strip() else len(line.split(sep))
+            if count != expected_count:
+                # could collect errors (first 10?) and throw an exception with a list
+                # lets wait and see if that's really needed
+                raise _ParseException(Error(
+                    ErrorType.PARSE_FAIL,
+                    f"Incorrect number of items in line {i + 2}, "
+                    + f"expected {expected_count}, got {count}",
+                    SpecificationSource(path)
+                ))
+
+def _process_dataframe(df: pandas.DataFrame, spec_source: SpecificationSource) -> ParseResult:
+    results = []
```

Comment:
It's kinda hard to tell what's going on in here 
So it's passing in a dataframe, and converts it to a dictionary but wait, actually it's not a dictionary, its a list of dictionaries. ( I had to test it here https://www.learnpython.org/en/Pandas_Basics )
We could make this more obvious  by doing

`records = df.to_dict(orient='records') # type: List`

and then
for record in records:
    for key in record:
    This part doesn't make sense. Isn't k[0] an index of a string?

I tried this code in https://www.learnpython.org/en/Pandas_Basics and I got a list of dicts with the key being a single character (Unless i copied and pasted this wrong)

Reply:
> So it's passing in a dataframe, and converts it to a dictionary but wait, actually it's not a dictionary, its a list of dictionaries.

Yeah, pandas packed a bit too much functionality into that method IMO.

I've switched things around and added comments, see what you think

Reply:
The type hint for that method must be horrendous, it returns like 5 different type structures

Reply:
Much better, thanks!